// @tenderly/hardhat-tenderly/dist/Tenderly.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tenderly = void 0;
const fs = __importStar(require("fs-extra"));
const path_1 = require("path");
const index_1 = require("./index");
const TenderlyService_1 = require("./tenderly/TenderlyService");
const TenderlyNetwork_1 = require("./TenderlyNetwork");
const util_1 = require("./util");
class Tenderly {
    constructor(hre) {
        this.env = hre;
        this.tenderlyNetwork = new TenderlyNetwork_1.TenderlyNetwork(hre);
    }
    async verify(...contracts) {
        const flatContracts = contracts.reduce((accumulator, value) => accumulator.concat(value), []);
        const requestData = await this.filterContracts(flatContracts);
        if (requestData == null) {
            console.log("Verification failed");
            return;
        }
        try {
            await TenderlyService_1.TenderlyService.verifyContracts(requestData);
        }
        catch (err) {
            console.log(err.message);
        }
    }
    network() {
        return this.tenderlyNetwork;
    }
    async push(...contracts) {
        const flatContracts = contracts.reduce((accumulator, value) => accumulator.concat(value), []);
        const requestData = await this.filterContracts(flatContracts);
        if (this.env.config.tenderly.project === undefined) {
            console.log(`Error in ${index_1.PluginName}: Please provide the project field in the tenderly object in hardhat.config.js`);
            return;
        }
        if (this.env.config.tenderly.username === undefined) {
            console.log(`Error in ${index_1.PluginName}: Please provide the username field in the tenderly object in hardhat.config.js`);
            return;
        }
        if (requestData == null) {
            console.log("Push failed");
            return;
        }
        try {
            await TenderlyService_1.TenderlyService.pushContracts(requestData, this.env.config.tenderly.project, this.env.config.tenderly.username);
        }
        catch (err) {
            console.log(err.message);
        }
    }
    async persistArtifacts(...contracts) {
        const sourcePaths = await this.env.run("compile:solidity:get-source-paths");
        const sourceNames = await this.env.run("compile:solidity:get-source-names", { sourcePaths });
        const data = await this.env.run("compile:solidity:get-dependency-graph", {
            sourceNames
        });
        let contract;
        const destPath = `deployments${path_1.sep}localhost_5777${path_1.sep}`;
        data._resolvedFiles.forEach((resolvedFile, _) => {
            const sourcePath = resolvedFile.sourceName;
            const name = sourcePath
                .split("/")
                .slice(-1)[0]
                .split(".")[0];
            for (contract of contracts) {
                if (contract.name === name) {
                    const contractDataPath = `${this.env.config.paths.artifacts}${path_1.sep}${sourcePath}${path_1.sep}${name}.json`;
                    const contractData = JSON.parse(fs.readFileSync(contractDataPath).toString());
                    const metadata = {
                        compiler: {
                            version: util_1.extractCompilerVersion(this.env.config, sourcePath)
                        },
                        sources: {
                            [sourcePath]: {
                                content: resolvedFile.content.rawContent
                            }
                        }
                    };
                    const visited = {};
                    util_1.resolveDependencies(data, sourcePath, metadata, visited);
                    const artifact = {
                        metadata: JSON.stringify(metadata),
                        address: contract.address,
                        bytecode: contractData.bytecode,
                        deployedBytecode: contractData.deployedBytecode,
                        abi: contractData.abi
                    };
                    fs.outputFileSync(`${destPath}${name}.json`, JSON.stringify(artifact));
                }
            }
        });
    }
    async filterContracts(flatContracts) {
        let contract;
        let requestData;
        try {
            requestData = await this.getContractData(flatContracts);
        }
        catch (e) {
            return null;
        }
        for (contract of flatContracts) {
            const network = this.env.hardhatArguments.network !== "hardhat"
                ? this.env.hardhatArguments.network || contract.network
                : contract.network;
            if (network === undefined) {
                console.log(`Error in ${index_1.PluginName}: Please provide a network via the hardhat --network argument or directly in the contract`);
                return null;
            }
            const index = requestData.contracts.findIndex(requestContract => requestContract.contractName === contract.name);
            if (index === -1) {
                continue;
            }
            requestData.contracts[index].networks = {
                [index_1.NetworkMap[network.toLowerCase()]]: {
                    address: contract.address
                }
            };
        }
        return requestData;
    }
    async getContractData(flatContracts) {
        const contracts = await util_1.getContracts(this.env, flatContracts);
        const config = util_1.getCompilerDataFromContracts(contracts, flatContracts, this.env.config);
        if (config === undefined) {
            console.log(`Error in ${index_1.PluginName}: No compiler configuration found`);
        }
        return {
            contracts,
            config: config
        };
    }
}
exports.Tenderly = Tenderly;
